flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPoolTest.java

```diff
@@ -0,0 +1,550 @@
+	}
+
+	@Test
+	public void testOfferingTooManySlots() {
+		final NewSlotsService notifyNewSlots = new NewSlotsService();
+		final DefaultDeclarativeSlotPool slotPool = createDefaultDeclarativeSlotPool(notifyNewSlots);
+
+		final ResourceCounter resourceRequirements = createResourceRequirements();
+
+		slotPool.increaseResourceRequirementsBy(resourceRequirements);
+
+		final ResourceCounter increasedRequirements = resourceRequirements.add(RESOURCE_PROFILE_1, 2);
+
+		final Collection<SlotOffer> slotOffers = createSlotOffersForResourceRequirements(increasedRequirements);
+
+		final Collection<SlotOffer> acceptedSlots = offerSlots(slotPool, slotOffers);
+
+		final Map<ResourceProfile, Long> resourceProfileCount = acceptedSlots.stream().map(SlotOffer::getResourceProfile).collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+
+		for (Map.Entry<ResourceProfile, Integer> resourceCount : resourceRequirements.getResourcesWithCount()) {
+			assertThat(resourceProfileCount.getOrDefault(resourceCount.getKey(), 0L), is((long) resourceCount.getValue()));
+		}
+	}
+
+	@Test
+	public void testReleaseSlotsRemovesSlots() throws InterruptedException {
+		final NewResourceRequirementsService notifyNewResourceRequirements = new NewResourceRequirementsService();
+		final DefaultDeclarativeSlotPool slotPool = createDefaultDeclarativeSlotPool(notifyNewResourceRequirements);
+
+		final LocalTaskManagerLocation taskManagerLocation = new LocalTaskManagerLocation();
+		increaseRequirementsAndOfferSlotsToSlotPool(slotPool, createResourceRequirements(), taskManagerLocation);
+
+		notifyNewResourceRequirements.takeResourceRequirements();
+
+		slotPool.releaseSlots(taskManagerLocation.getResourceID(), new FlinkException("Test failure"));
+		assertThat(slotPool.getAllSlotsInformation(), is(empty()));
+	}
+
+	@Test
+	public void testReleaseSlotsReturnsSlot() {
+		final DefaultDeclarativeSlotPool slotPool = DefaultDeclarativeSlotPoolBuilder.builder().build();
+
+		final ResourceCounter resourceRequirements = createResourceRequirements();
+
+		final LocalTaskManagerLocation taskManagerLocation = new LocalTaskManagerLocation();
+		final FreeSlotConsumer freeSlotConsumer = new FreeSlotConsumer();
+		final TestingTaskExecutorGateway testingTaskExecutorGateway = new TestingTaskExecutorGatewayBuilder()
+			.setFreeSlotFunction(freeSlotConsumer)
+			.createTestingTaskExecutorGateway();
+
+		final Collection<SlotOffer> slotOffers = increaseRequirementsAndOfferSlotsToSlotPool(
+			slotPool,
+			resourceRequirements,
+			taskManagerLocation,
+			testingTaskExecutorGateway);
+
+		slotPool.releaseSlots(taskManagerLocation.getResourceID(), new FlinkException("Test failure"));
+
+		final Collection<AllocationID> freedSlots = freeSlotConsumer.drainFreedSlots();
+
+		assertThat(freedSlots, containsInAnyOrder(slotOffers.stream().map(SlotOffer::getAllocationId).toArray()));
+	}
+
+	@Test
+	public void testFailSlotDecreasesResources() throws InterruptedException {
```

Comment:
Maybe better to call this test `testReleaseSlotDecreasesFulfilledResourceRequirements`.